--- binutils-2.22.52.0.4/bfd/bfd-in2.h.orig	2012-06-14 16:29:22.000000000 +0200
+++ binutils-2.22.52.0.4/bfd/bfd-in2.h	2012-06-14 20:26:46.835857525 +0200
@@ -32,11 +32,6 @@
 #ifndef __BFD_H_SEEN__
 #define __BFD_H_SEEN__
 
-/* PR 14072: Ensure that config.h is included first.  */
-#if !defined PACKAGE && !defined PACKAGE_VERSION
-#error config.h must be included before this header
-#endif
-
 #ifdef __cplusplus
 extern "C" {
 #endif
@@ -39,6 +39,7 @@
 #include <stdarg.h>
 #include <string.h>
 #include <sys/stat.h>
+#include <stddef.h>
 
 #if defined (__STDC__) || defined (ALMOST_STDC) || defined (HAVE_STRINGIZE)
 #ifndef SABER
