--- binutils-2.23.51.0.5/bfd/compress.c.orig	2012-11-12 17:45:23.000000000 +0100
+++ binutils-2.23.51.0.5/bfd/compress.c	2012-11-13 17:20:33.925842054 +0100
@@ -178,7 +178,7 @@ bfd_get_full_section_contents (bfd *abfd
     case COMPRESS_SECTION_NONE:
       if (p == NULL)
 	{
-	  p = (bfd_byte *) bfd_malloc (sz);
+	  p = (bfd_byte *) bfd_zmalloc (sz);
 	  if (p == NULL)
 	    return FALSE;
 	}
@@ -218,7 +218,7 @@ bfd_get_full_section_contents (bfd *abfd
 	goto fail_compressed;
 
       if (p == NULL)
-	p = (bfd_byte *) bfd_malloc (sz);
+	p = (bfd_byte *) bfd_zmalloc (sz);
       if (p == NULL)
 	goto fail_compressed;
 
