--- binutils-2.28/bfd/elflink.c.orig	2017-06-05 22:10:50.459268960 +0200
+++ binutils-2.28/bfd/elflink.c	2017-06-05 22:30:10.125922386 +0200
@@ -10420,7 +10420,7 @@
 		      if (action_discarded & COMPLAIN)
 			(*flinfo->info->callbacks->einfo)
 			  /* xgettext:c-format */
-			  (_("%X`%s' referenced in section `%A' of %B: "
+			  (_("`%s' referenced in section `%A' of %B: "
 			     "defined in discarded section `%A' of %B\n"),
 			   sym_name, o, input_bfd, sec, sec->owner);
 
