--- binutils-2.21.51.0.1/bfd/doc/bfd.texinfo.orig	2010-11-12 18:32:59.000000000 +0100
+++ binutils-2.21.51.0.1/bfd/doc/bfd.texinfo	2010-11-20 11:27:17.171103290 +0100
@@ -7,9 +7,9 @@
 @synindex fn cp
 
 @ifnottex
-@dircategory Software development
+@dircategory Programming tools:
 @direntry
-* Bfd: (bfd).                   The Binary File Descriptor library.
+* Bfd: (bfd).				The Binary File Descriptor library
 @end direntry
 @end ifnottex
 
diff -urN binutils-2.12.90.0.1.org/etc/configure.texi binutils-2.12.90.0.1/etc/configure.texi
--- binutils-2.12.90.0.1.org/etc/configure.texi	Mon Mar 11 22:37:37 2002
+++ binutils-2.12.90.0.1/etc/configure.texi	Mon Mar 11 22:37:52 2002
@@ -5,9 +5,9 @@
 @setchapternewpage off
 @c %**end of header
 
-@dircategory GNU admin
+@dircategory Programming tools:
 @direntry
-* configure: (configure).	The GNU configure and build system
+* configure: (configure).		The GNU configure and build system
 @end direntry
 
 @ifinfo
--- binutils-2.20.51.0.11/etc/standards.texi.orig	2010-07-09 21:35:45.000000000 +0200
+++ binutils-2.20.51.0.11/etc/standards.texi	2010-08-12 08:16:46.230823295 +0200
@@ -6,9 +6,9 @@
 @set lastupdate April 12, 2010
 @c %**end of header
 
-@dircategory GNU organization
+@dircategory Miscellaneous:
 @direntry
-* Standards: (standards).       GNU coding standards.
+* Standards: (standards).		GNU coding standards
 @end direntry
 
 @c @setchapternewpage odd
--- binutils-2.21.51.0.1/gas/doc/as.texinfo.orig	2010-11-12 18:33:04.000000000 +0100
+++ binutils-2.21.51.0.1/gas/doc/as.texinfo	2010-11-20 11:27:52.647104966 +0100
@@ -88,10 +88,10 @@
 @end ifinfo
 
 @ifnottex
-@dircategory Software development
+@dircategory Programming tools:
 @direntry
-* As: (as).                     The GNU assembler.
-* Gas: (as).                    The GNU assembler.
+* As: (as).				The GNU assembler
+* Gas: (as).				The GNU assembler
 @end direntry
 @end ifnottex
 
diff -urN binutils-2.14.90.0.7.org/gas/doc/gasp.texinfo binutils-2.14.90.0.7/gas/doc/gasp.texinfo
--- binutils-2.14.90.0.7.org/gas/doc/gasp.texinfo	Mon Mar 11 22:37:38 2002
+++ binutils-2.14.90.0.7/gas/doc/gasp.texinfo	Mon Mar 11 22:37:52 2002
@@ -14,9 +14,10 @@
 
 @ifinfo
 @format
-START-INFO-DIR-ENTRY
-* gasp: (gasp).                     The GNU Assembler Preprocessor
-END-INFO-DIR-ENTRY
+@dircategory Programming tools:
+@direntry
+* gasp: (gasp).				The GNU Assembler Preprocessor
+@end direntry
 @end format
 @end ifinfo
 
--- binutils-2.21.51.0.1/gprof/gprof.texi.orig	2010-11-12 18:33:07.000000000 +0100
+++ binutils-2.21.51.0.1/gprof/gprof.texi	2010-11-20 11:28:30.103104128 +0100
@@ -13,9 +13,9 @@
 @ifnottex
 @c This is a dir.info fragment to support semi-automated addition of
 @c manuals to an info tree.  zoo@cygnus.com is developing this facility.
-@dircategory Software development
+@dircategory Programming tools:
 @direntry
-* gprof: (gprof).                Profiling your program's execution
+* gprof: (gprof).			Profiling your program's execution
 @end direntry
 @end ifnottex
 
--- binutils-2.21.51.0.1/ld/ld.texinfo.orig	2010-11-12 18:33:07.000000000 +0100
+++ binutils-2.21.51.0.1/ld/ld.texinfo	2010-11-20 11:28:52.127099099 +0100
@@ -40,9 +40,9 @@
 @c man end
 
 @ifnottex
-@dircategory Software development
+@dircategory Programming tools:
 @direntry
-* Ld: (ld).                       The GNU linker.
+* Ld: (ld).				The GNU linker
 @end direntry
 @end ifnottex
 
@@ -7320,7 +7320,7 @@
 @node BFD outline
 @section How It Works: An Outline of BFD
 @cindex opening object files
-@include bfdsumm.texi
+@include ../bfd/doc/bfdsumm.texi
 @end ifclear
 
 @node Reporting Bugs
--- binutils-2.18.50.0.5/binutils/doc/binutils.texi.orig	2008-03-14 19:23:02.000000000 +0100
+++ binutils-2.18.50.0.5/binutils/doc/binutils.texi	2008-03-18 08:27:03.340177183 +0100
@@ -23,30 +23,26 @@
 @c man end
 @end copying
 
-@dircategory Software development
+@dircategory Programming tools:
 @direntry
-* Binutils: (binutils).         The GNU binary utilities.
-@end direntry
-
-@dircategory Individual utilities
-@direntry
-* addr2line: (binutils)addr2line. Convert addresses to file and line.
-* ar: (binutils)ar.               Create, modify, and extract from archives.
-* c++filt: (binutils)c++filt.	  Filter to demangle encoded C++ symbols.
-* cxxfilt: (binutils)c++filt.     MS-DOS name for c++filt.
-* dlltool: (binutils)dlltool.	  Create files needed to build and use DLLs.
-* nlmconv: (binutils)nlmconv.     Converts object code into an NLM.
-* nm: (binutils)nm.               List symbols from object files.
-* objcopy: (binutils)objcopy.	  Copy and translate object files.
-* objdump: (binutils)objdump.     Display information from object files.
-* ranlib: (binutils)ranlib.       Generate index to archive contents.
-* readelf: (binutils)readelf.	  Display the contents of ELF format files.
-* size: (binutils)size.           List section sizes and total size.
-* strings: (binutils)strings.     List printable strings from files.
-* strip: (binutils)strip.         Discard symbols.
-* elfedit: (binutils)elfedit.     Update the ELF header of ELF files.
-* windmc: (binutils)windmc.	  Generator for Windows message resources.
-* windres: (binutils)windres.	  Manipulate Windows resources.
+* Binutils: (binutils).			The GNU binary utilities
+* addr2line: (binutils)addr2line.	Convert addresses to file and line
+* ar: (binutils)ar.			Create, modify, and extract from archives
+* c++filt: (binutils)c++filt.		Filter to demangle encoded C++ symbols
+* cxxfilt: (binutils)c++filt.		MS-DOS name for c++filt
+* dlltool: (binutils)dlltool.		Create files needed to build and use DLLs
+* nlmconv: (binutils)nlmconv.		Converts object code into an NLM
+* nm: (binutils)nm.			List symbols from object files
+* objcopy: (binutils)objcopy.		Copy and translate object files
+* objdump: (binutils)objdump.		Display information from object files
+* ranlib: (binutils)ranlib.		Generate index to archive contents
+* readelf: (binutils)readelf.		Display the contents of ELF format files
+* size: (binutils)size.			List section sizes and total size
+* strings: (binutils)strings.		List printable strings from files
+* strip: (binutils)strip.		Discard symbols
+* elfedit: (binutils)elfedit.		Update the ELF header of ELF files
+* windmc: (binutils)windmc.		Generator for Windows message resources
+* windres: (binutils)windres.		Manipulate Windows resources
 @end direntry
 
 @titlepage
