diff -urN binutils-2.11.90.0.7.org/bfd/linker.c binutils-2.11.90.0.7/bfd/linker.c
--- binutils-2.11.90.0.7.org/bfd/linker.c	Thu May 10 12:26:36 2001
+++ binutils-2.11.90.0.7/bfd/linker.c	Thu May 10 12:26:49 2001
@@ -1801,6 +1801,15 @@
 						copy, false);
 	    if (inh == (struct bfd_link_hash_entry *) NULL)
 	      return false;
+	    if (inh->u.i.link == h
+		&& inh->type == bfd_link_hash_indirect)
+	      {
+		(*_bfd_error_handler)
+		  (_("%s: indirect symbol `%s' to `%s' is a loop"),  
+		   bfd_get_filename (abfd), name, string);
+		bfd_set_error (bfd_error_invalid_operation);
+		return false;
+	      }
 	    if (inh->type == bfd_link_hash_new)
 	      {
 		inh->type = bfd_link_hash_undefined;
