--- binutils-2.13.90.0.16/gas/Makefile.am.orig	Thu Nov 14 18:37:56 2002
+++ binutils-2.13.90.0.16/gas/Makefile.am	Thu Jan 30 18:32:07 2003
@@ -449,7 +449,7 @@
 
 # Note: GASP is now deprecated and will be removed at some point in the future.
 # Anything that GASP could do can now be done by GAS.
-noinst_PROGRAMS = as-new
+noinst_PROGRAMS = as-new gasp-new
 noinst_SCRIPTS = $(GDBINIT)
 EXTRA_SCRIPTS = .gdbinit
 
@@ -648,7 +648,7 @@
 
 # stand-alone itbl assembler & disassembler
 
-EXTRA_PROGRAMS = gasp-new itbl-test
+EXTRA_PROGRAMS = itbl-test
 itbl_test_SOURCES = itbl-parse.y itbl-lex.l
 itbl_test_LDADD = itbl-tops.o itbl-test.o $(GASLIBS) @LEXLIB@
 
--- binutils-2.13.90.0.16/gas/doc/Makefile.am.orig	Mon Jul 29 17:19:18 2002
+++ binutils-2.13.90.0.16/gas/doc/Makefile.am	Thu Jan 30 18:34:31 2003
@@ -17,7 +17,7 @@
 
 # The gasp.texi file is now deprecated, as GASP has been deprecated.
 # It will disappear in a future release.
-info_TEXINFOS = as.texinfo 
+info_TEXINFOS = as.texinfo gasp.texi
 
 asconfig.texi: $(CONFIG).texi
 	rm -f asconfig.texi
@@ -71,7 +71,7 @@
 
 # This one isn't ready for prime time yet.  Not even a little bit.
 
-noinst_TEXINFOS = internals.texi gasp.texi
+noinst_TEXINFOS = internals.texi
 
 DISTCLEANFILES = asconfig.texi
 
--- binutils-2.13.90.0.16/gas/Makefile.in.orig	Thu Nov 14 18:37:56 2002
+++ binutils-2.13.90.0.16/gas/Makefile.in	Thu Jan 30 18:33:00 2003
@@ -572,7 +572,7 @@
 
 # Note: GASP is now deprecated and will be removed at some point in the future.
 # Anything that GASP could do can now be done by GAS.
-noinst_PROGRAMS = as-new
+noinst_PROGRAMS = as-new gasp-new
 noinst_SCRIPTS = $(GDBINIT)
 EXTRA_SCRIPTS = .gdbinit
 
@@ -642,7 +642,7 @@
 
 # stand-alone itbl assembler & disassembler
 
-EXTRA_PROGRAMS = gasp-new itbl-test
+EXTRA_PROGRAMS = itbl-test
 itbl_test_SOURCES = itbl-parse.y itbl-lex.l
 itbl_test_LDADD = itbl-tops.o itbl-test.o $(GASLIBS) @LEXLIB@
 
@@ -2372,7 +2372,7 @@
 mkinstalldirs = $(SHELL) $(top_srcdir)/../mkinstalldirs
 CONFIG_HEADER = config.h
 CONFIG_CLEAN_FILES =  ${GDBINIT}
-noinst_PROGRAMS =  as-new$(EXEEXT)
+noinst_PROGRAMS =  as-new$(EXEEXT) gasp-new$(EXEEXT)
 PROGRAMS =  $(noinst_PROGRAMS)
 
 
--- binutils-2.13.90.0.16/gas/doc/Makefile.in.orig	Wed Oct 30 05:09:13 2002
+++ binutils-2.13.90.0.16/gas/doc/Makefile.in	Thu Jan 30 18:35:54 2003
@@ -135,7 +135,7 @@
 
 # The gasp.texi file is now deprecated, as GASP has been deprecated.
 # It will disappear in a future release.
-info_TEXINFOS = as.texinfo 
+info_TEXINFOS = as.texinfo gasp.texi
 
 CPU_DOCS = \
 	c-a29k.texi \
@@ -173,7 +173,7 @@
 
 # This one isn't ready for prime time yet.  Not even a little bit.
 
-noinst_TEXINFOS = internals.texi gasp.texi
+noinst_TEXINFOS = internals.texi
 
 DISTCLEANFILES = asconfig.texi
 
@@ -183,9 +183,9 @@
 CONFIG_CLEAN_FILES = 
 TEXI2DVI = `if test -f $(top_srcdir)/../texinfo/util/texi2dvi; then echo $(top_srcdir)/../texinfo/util/texi2dvi; else echo texi2dvi; fi`
 TEXINFO_TEX = $(top_srcdir)/../texinfo/texinfo.tex
-INFO_DEPS = as.info
+INFO_DEPS = as.info gasp.info
 DVIS = as.dvi
-TEXINFOS = as.texinfo 
+TEXINFOS = as.texinfo gasp.texi
 man1dir = $(mandir)/man1
 MANS = $(man_MANS)
 
